src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -1396,23 +1503,28 @@ private void doEncodeBegin(
             int remoteChannel,
             int nextOutgoingId)
         {
-            final AmqpBeginFW begin = amqpBeginRW.wrap(frameBuffer, FRAME_HEADER_SIZE, frameBuffer.capacity())
+            final AmqpBeginFW begin = amqpBeginRW.wrap(frameBuffer, FRAME_HEADER_SIZE + PERFORMATIVE_SIZE, frameBuffer.capacity())
                 .remoteChannel(remoteChannel)
                 .nextOutgoingId(nextOutgoingId)
                 .incomingWindow(bufferPool.slotCapacity())
                 .outgoingWindow(outgoingWindow)
                 .build();
 
+            final int size = FRAME_HEADER_SIZE + PERFORMATIVE_SIZE + begin.sizeof();
+
             final AmqpFrameHeaderFW frameHeader = amqpFrameHeaderRW.wrap(frameBuffer, 0, frameBuffer.capacity())
-                .size(FRAME_HEADER_SIZE + begin.sizeof())
+                .size(size)
                 .doff(2)
                 .type(0)
                 .channel(outgoingChannel)
-                .performative(b -> b.begin(begin))
                 .build();
 
-            replyBudgetReserved += frameHeader.sizeof() + replyPadding;
-            doNetworkData(traceId, authorization, 0L, frameHeader);
+            frameBuffer.putBytes(frameHeader.limit(), beginType.buffer(), 0, PERFORMATIVE_SIZE);
```

Comment:
This should really be `beginType.sizeof()` instead of `PERFORMATIVE_SIZE`.

Suggest reordering this to put the performative first using `FRAME_HEADER_SIZE`, so that it reads as if you are building up the entire frame body, then using that information to fill in the size field of the frame header.

Then after filling in the frame header, add an `assert frameHeader.sizeof() == FRAME_HEADER_SIZE` to ensure your assumptions about positional layout are either always correct, or detected as inconsistent by the assert, perhaps in future if the code is modified.

Please apply this feedback to each of the decodeXXX methods for performatives.